File: cli-go/internal/parser/parser.go
package parser

import (
	"regexp"
	"strconv"
	"strings"
)

// StereoMode describes the 3-D layout of the video frame.
type StereoMode string

const (
	StereoSBS StereoMode = "sbs"
	StereoTB  StereoMode = "tb"
	StereoOff StereoMode = "off"
)

// Projection describes the lens / equirect geometry.
type Projection string

const (
	ProjEquirect180 Projection = "equirect-180"
	ProjEquirect360 Projection = "equirect-360"
	ProjMKX200      Projection = "mkx200"
	ProjMKX220      Projection = "mkx220"
	ProjRF52        Projection = "rf52"
	ProjFisheye     Projection = "fisheye"
	ProjFlat        Projection = "flat"
)

// ScreenTypeForDeovr maps Projection to DeoVR's screenType field.
var ScreenTypeForDeovr = map[Projection]string{
	ProjEquirect180: "dome",
	ProjEquirect360: "sphere",
	ProjMKX200:      "mkx200",
	ProjMKX220:      "mkx220",
	ProjRF52:        "rf52",
	ProjFisheye:     "fisheye",
	ProjFlat:        "flat",
}

// VrMetadata is the structured result of parsing a VR video filename.
type VrMetadata struct {
	Filename        string     `json:"filename"`
	Title           string     `json:"title"`
	Studio          *string    `json:"studio"`
	Resolution      *int       `json:"resolution"`
	ResolutionLabel *string    `json:"resolutionLabel"`
	Projection      Projection `json:"projection"`
	FOV             int        `json:"fov"`
	StereoMode      StereoMode `json:"stereoMode"`
	Is3D            bool       `json:"is3d"`
}

type studioMatcher struct {
	studio  string
	pattern *regexp.Regexp
}

var studioMatchers = []studioMatcher{
	{studio: "NaughtyAmericaVR", pattern: regexp.MustCompile(`(?i)^NaughtyAmericaVR\b`)},
	{studio: "VirtualRealPorn", pattern: regexp.MustCompile(`(?i)^VirtualRealPorn[_-]`)},
	{studio: "VRBangers", pattern: regexp.MustCompile(`(?i)^VRBANGERS[_-]`)},
	{studio: "WankzVR", pattern: regexp.MustCompile(`(?i)^wankzvr[_-]`)},
	{studio: "SLR", pattern: regexp.MustCompile(`(?i)^SLR[_-]`)},
	{studio: "Manny S (SLR)", pattern: regexp.MustCompile(`(?i)^Manny_S[_-]`)},
}

// Short codes handled separately because Go RE2 doesn't support lookahead.
var naShortCodes = []string{"nam", "naw", "tdrm", "tspa", "ptgs"}

var tagTokenSet = map[string]struct{}{
	"mkx200": {}, "mkx220": {}, "rf52": {}, "fisheye": {},
	"180x180": {}, "360x180": {}, "180": {}, "360": {},
	"lr": {}, "tb": {}, "ou": {}, "sbs": {}, "3dh": {}, "3dv": {}, "mono": {},
	"vr265": {}, "vrdesktophd": {}, "8kvr265": {}, "6kvr265": {}, "4kvr265": {},
	"2k": {}, "4k": {}, "6k": {}, "8k": {},
}

func isTagToken(token string) bool {
	lower := strings.ToLower(token)
	if _, ok := tagTokenSet[lower]; ok {
		return true
	}
	// \d{3,4}p  e.g. 2900p, 1080p, 4096p
	if len(lower) >= 2 && lower[len(lower)-1] == 'p' {
		num := lower[:len(lower)-1]
		if len(num) >= 3 && len(num) <= 4 {
			allDigits := true
			for _, c := range num {
				if c < '0' || c > '9' {
					allDigits = false
					break
				}
			}
			if allDigits {
				return true
			}
		}
	}
	// numeric 4-6 digits (scene IDs)
	if len(lower) >= 4 && len(lower) <= 6 {
		allDigits := true
		for _, c := range lower {
			if c < '0' || c > '9' {
				allDigits = false
				break
			}
		}
		if allDigits {
			return true
		}
	}
	return false
}

var (
	rePx        = regexp.MustCompile(`(?i)(\d{3,4})p`)
	re8kvr265   = regexp.MustCompile(`(?i)8kvr265|_8K(_|\b)|8K_`)
	re6kvr265   = regexp.MustCompile(`(?i)6kvr265|_6K(_|\b)|6K_`)
	re4kvr265   = regexp.MustCompile(`(?i)_4K(_|\b)|4K_|4kvr265`)
	re2kvr265   = regexp.MustCompile(`(?i)_2K(_|\b)|2K_`)
	reVRDesktop = regexp.MustCompile(`(?i)vrdesktophd`)

	reMKX200  = regexp.MustCompile(`(?i)MKX200`)
	reMKX220  = regexp.MustCompile(`(?i)MKX220`)
	reRF52    = regexp.MustCompile(`(?i)RF52`)
	reFisheye = regexp.MustCompile(`(?i)FISHEYE`)
	re360     = regexp.MustCompile(`(?i)(?:^|[_\-])360(?:[_\-.]|$)|360x180`)
	reMono    = regexp.MustCompile(`(?i)(?:^|[_\-])mono(?:[_\-.]|$)|(?:^|[_\-])2D(?:[_\-.]|$)`)
	reTB      = regexp.MustCompile(`(?i)(?:^|[_\-])(?:TB|OU|3dv)(?:[_\-.]|$)`)

	reExt       = regexp.MustCompile(`(?i)\.[a-z0-9]+$`)
	reSplit     = regexp.MustCompile(`[_]+|\s*-\s*`)
	reTrimLead  = regexp.MustCompile(`^[_\-\s]+`)
	reSpace     = regexp.MustCompile(`\s+`)
	reUnderscore = regexp.MustCompile(`[_]+`)
)

func detectResolution(filename string) (*int, *string) {
	if m := rePx.FindStringSubmatch(filename); m != nil {
		n, _ := strconv.Atoi(m[1])
		label := strconv.Itoa(n) + "p"
		return &n, &label
	}
	if re8kvr265.MatchString(filename) {
		n := 4320
		l := "8K"
		return &n, &l
	}
	if re6kvr265.MatchString(filename) {
		n := 3160
		l := "6K"
		return &n, &l
	}
	if re4kvr265.MatchString(filename) {
		n := 2160
		l := "4K"
		return &n, &l
	}
	if re2kvr265.MatchString(filename) {
		n := 1440
		l := "2K"
		return &n, &l
	}
	if reVRDesktop.MatchString(filename) {
		n := 1080
		l := "HD"
		return &n, &l
	}
	return nil, nil
}

func detectProjection(filename string) (Projection, int) {
	if reMKX200.MatchString(filename) {
		return ProjMKX200, 200
	}
	if reMKX220.MatchString(filename) {
		return ProjMKX220, 220
	}
	if reRF52.MatchString(filename) {
		return ProjRF52, 190
	}
	if reFisheye.MatchString(filename) {
		return ProjFisheye, 180
	}
	if re360.MatchString(filename) {
		return ProjEquirect360, 360
	}
	return ProjEquirect180, 180
}

func detectStereoMode(filename string) StereoMode {
	if reMono.MatchString(filename) {
		return StereoOff
	}
	if reTB.MatchString(filename) {
		return StereoTB
	}
	return StereoSBS
}

func detectStudio(stem string) (*string, string) {
	for _, m := range studioMatchers {
		loc := m.pattern.FindStringIndex(stem)
		if loc != nil && loc[0] == 0 {
			matched := stem[loc[0]:loc[1]]
			remainder := stem[len(matched):]
			remainder = reTrimLead.ReplaceAllString(remainder, "")
			s := m.studio
			return &s, remainder
		}
	}
	// Check short codes: prefix + lookahead [a-z] (case-insensitive).
	lower := strings.ToLower(stem)
	for _, code := range naShortCodes {
		if strings.HasPrefix(lower, code) && len(stem) > len(code) {
			next := stem[len(code)]
			if (next >= 'a' && next <= 'z') || (next >= 'A' && next <= 'Z') {
				remainder := stem[len(code):]
				remainder = reTrimLead.ReplaceAllString(remainder, "")
				s := "NaughtyAmerica"
				return &s, remainder
			}
		}
	}
	return nil, stem
}

func extractTitle(remainder, fullStem string) string {
	if remainder == "" {
		return fullStem
	}
	tokens := reSplit.Split(remainder, -1)
	var titleTokens []string
	for _, t := range tokens {
		if t == "" {
			continue
		}
		if !isTagToken(t) {
			titleTokens = append(titleTokens, t)
		}
	}
	if len(titleTokens) == 0 {
		fallback := reUnderscore.ReplaceAllString(remainder, " ")
		fallback = strings.TrimSpace(fallback)
		if fallback != "" {
			return fallback
		}
		return fullStem
	}
	joined := strings.Join(titleTokens, " ")
	joined = reSpace.ReplaceAllString(joined, " ")
	return strings.TrimSpace(joined)
}

// ParseVrFilename parses a VR video filename into structured metadata.
// Example: "Sample_Title_2900p_MKX200.mp4"
func ParseVrFilename(filename string) VrMetadata {
	stem := reExt.ReplaceAllString(filename, "")

	studio, remainder := detectStudio(stem)
	resolution, label := detectResolution(filename)
	projection, fov := detectProjection(filename)
	stereoMode := detectStereoMode(filename)
	title := extractTitle(remainder, stem)

	return VrMetadata{
		Filename:        filename,
		Title:           title,
		Studio:          studio,
		Resolution:      resolution,
		ResolutionLabel: label,
		Projection:      projection,
		FOV:             fov,
		StereoMode:      stereoMode,
		Is3D:            stereoMode != StereoOff,
	}
}

File: shared/src/parser.ts
/**
 * Parse VR-video filenames into structured metadata.
 *
 * Filename conventions encode:
 * - Studio prefix (e.g. "STUDIO_A_", "STUDIO-B-")
 * - Resolution (e.g. "2900p", "4096p", "8K", "vrdesktophd")
 * - Projection (e.g. "MKX200", "180x180", "fisheye")
 * - Stereo mode (e.g. "_LR", "_TB", "3dh")
 *
 * Output is best-effort: missing fields are returned null. Title falls back to
 * the cleaned filename minus extension if no studio is detected.
 *
 * Used by both worker (catalog rendering) and CLI (R2 key + thumb generation).
 */
export type StereoMode = 'sbs' | 'tb' | 'off';

export type Projection =
  | 'equirect-180'
  | 'equirect-360'
  | 'mkx200'
  | 'mkx220'
  | 'rf52'
  | 'fisheye'
  | 'flat';

/** Maps to DeoVR's `screenType` field. */
export const screenTypeForDeovr: Record<Projection, string> = {
  'equirect-180': 'dome',
  'equirect-360': 'sphere',
  mkx200: 'mkx200',
  mkx220: 'mkx220',
  rf52: 'rf52',
  fisheye: 'fisheye',
  flat: 'flat',
};

export interface VrMetadata {
  /** Original filename including extension. R2 object key suffix. */
  filename: string;
  /** Best-effort human-friendly title. */
  title: string;
  /** Detected studio (canonical short name) or null. */
  studio: string | null;
  /** Vertical resolution in pixels, e.g. 2900, 4096, 4320. Null if unknown. */
  resolution: number | null;
  /** Display label like "8K", "2900p", "1080p". */
  resolutionLabel: string | null;
  /** Projection geometry. Defaults to equirect-180 (most common in VR). */
  projection: Projection;
  /** Field of view in degrees, derived from projection. */
  fov: number;
  /** Stereo mode. Defaults to sbs (most VR files are SBS). */
  stereoMode: StereoMode;
  /** True if not flat/mono. */
  is3d: boolean;
}

interface StudioMatcher {
  /** Canonical short label. */
  studio: string;
  /** Detection regex; must match start of filename. */
  pattern: RegExp;
}

/**
 * Studio detection. Order matters — more specific patterns first.
 * Short prefix codes handled last so longer matches win.
 */
const STUDIO_MATCHERS: StudioMatcher[] = [
  { studio: 'NaughtyAmericaVR', pattern: /^NaughtyAmericaVR\b/i },
  { studio: 'VirtualRealPorn', pattern: /^VirtualRealPorn[_-]/i },
  { studio: 'VRBangers', pattern: /^VRBANGERS[_-]/i },
  { studio: 'WankzVR', pattern: /^wankzvr[_-]/i },
  { studio: 'SLR', pattern: /^SLR[_-]/i },
  { studio: 'Manny S (SLR)', pattern: /^Manny_S[_-]/i },
  // NaughtyAmerica internal short codes — keep last so longer matches win.
  // Lookahead so we don't consume the first letter of the title fragment.
  { studio: 'NaughtyAmerica', pattern: /^(?:nam|naw|tdrm|tspa|ptgs)(?=[a-z])/i },
];

/** Tag tokens that are NOT part of the title — used to trim noise. */
const TAG_TOKENS = [
  // Resolution
  /^\d{3,4}p$/i,
  /^[2468]K$/i,
  /^4096p$/,
  // Projection
  /^MKX200$/i,
  /^MKX220$/i,
  /^RF52$/i,
  /^FISHEYE$/i,
  /^180x180$/i,
  /^360x180$/i,
  /^180$/,
  /^360$/,
  // Stereo
  /^LR$/i,
  /^TB$/i,
  /^OU$/i,
  /^SBS$/i,
  /^3dh$/i,
  /^3dv$/i,
  /^mono$/i,
  // Codecs / containers / generic
  /^vr265$/i,
  /^vrdesktophd$/i,
  /^8kvr265$/i,
  /^6kvr265$/i,
  /^4kvr265$/i,
  // numeric scene IDs (e.g. 31956)
  /^\d{4,6}$/,
];

function isTagToken(token: string): boolean {
  return TAG_TOKENS.some((re) => re.test(token));
}

/** Extract resolution number + label from filename. */
function detectResolution(filename: string): { resolution: number | null; label: string | null } {
  // Numeric "{num}p" wins over keyword "8K" since it's more specific.
  const px = filename.match(/(\d{3,4})p/i);
  if (px) {
    const n = parseInt(px[1], 10);
    return { resolution: n, label: `${n}p` };
  }

  // VR keyword shortcuts. Approximate vertical pixels.
  if (/8kvr265|_8K(_|\b)|8K_/i.test(filename)) return { resolution: 4320, label: '8K' };
  if (/6kvr265|_6K(_|\b)|6K_/i.test(filename)) return { resolution: 3160, label: '6K' };
  if (/_4K(_|\b)|4K_|4kvr265/i.test(filename)) return { resolution: 2160, label: '4K' };
  if (/_2K(_|\b)|2K_/i.test(filename)) return { resolution: 1440, label: '2K' };
  if (/vrdesktophd/i.test(filename)) return { resolution: 1080, label: 'HD' };

  return { resolution: null, label: null };
}

/** Detect projection geometry from filename keywords. */
function detectProjection(filename: string): { projection: Projection; fov: number } {
  if (/MKX200/i.test(filename)) return { projection: 'mkx200', fov: 200 };
  if (/MKX220/i.test(filename)) return { projection: 'mkx220', fov: 220 };
  if (/RF52/i.test(filename)) return { projection: 'rf52', fov: 190 };
  if (/FISHEYE/i.test(filename)) return { projection: 'fisheye', fov: 180 };
  if (/(?:^|[_\-])360(?:[_\-.]|$)|360x180/i.test(filename)) {
    return { projection: 'equirect-360', fov: 360 };
  }
  // Default for VR content: equirect-180.
  return { projection: 'equirect-180', fov: 180 };
}

/** Detect stereo layout from filename hints. */
function detectStereoMode(filename: string): StereoMode {
  if (/(?:^|[_\-])mono(?:[_\-.]|$)|(?:^|[_\-])2D(?:[_\-.]|$)/i.test(filename)) return 'off';
  if (/(?:^|[_\-])(?:TB|OU|3dv)(?:[_\-.]|$)/i.test(filename)) return 'tb';
  // sbs is the safe default for VR content; LR/3dh confirms it.
  return 'sbs';
}

/** Detect studio prefix; return matcher and the trimmed remainder. */
function detectStudio(stem: string): { studio: string | null; remainder: string } {
  for (const m of STUDIO_MATCHERS) {
    const match = stem.match(m.pattern);
    if (match) {
      const remainder = stem.slice(match[0].length).replace(/^[_\-\s]+/, '');
      return { studio: m.studio, remainder };
    }
  }
  return { studio: null, remainder: stem };
}

/**
 * Extract a human-readable title from the filename remainder (post-studio).
 *
 * Strategy: split on `_` and `-`, drop tag tokens (resolution, projection,
 * stereo), join the rest with spaces. If everything looks like tags or short
 * codes, return the original cleaned stem so we always have something.
 */
function extractTitle(remainder: string, fullStem: string): string {
  if (!remainder) return fullStem;

  // Split on underscores or runs of dashes/spaces, but preserve internal spaces.
  const tokens = remainder.split(/[_]+|\s*-\s*/).filter(Boolean);
  const titleTokens = tokens.filter((t) => !isTagToken(t));

  if (titleTokens.length === 0) {
    // Everything was a tag — fall back to the un-trimmed remainder.
    return remainder.replace(/[_]+/g, ' ').trim() || fullStem;
  }

  return titleTokens.join(' ').replace(/\s+/g, ' ').trim();
}

/**
 * Parse a VR video filename into structured metadata.
 *
 * @param filename Object key suffix, e.g. "Sample_Title_2900p_MKX200.mp4"
 */
export function parseVrFilename(filename: string): VrMetadata {
  // Strip extension for analysis but keep the original for the `filename` field.
  const stem = filename.replace(/\.[a-z0-9]+$/i, '');

  const { studio, remainder } = detectStudio(stem);
  const { resolution, label } = detectResolution(filename);
  const { projection, fov } = detectProjection(filename);
  const stereoMode = detectStereoMode(filename);
  const title = extractTitle(remainder, stem);

  return {
    filename,
    title,
    studio,
    resolution,
    resolutionLabel: label,
    projection,
    fov,
    stereoMode,
    is3d: stereoMode !== 'off',
  };
}
